// @signalapp/libsignal-client/dist/net.js
"use strict";
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Net = exports.SIGNAL_TLS_PROXY_SCHEME = exports.TokioAsyncContext = exports.newNativeHandle = exports.Environment = void 0;
const Native = require("../Native");
const CDSI_1 = require("./net/CDSI");
const Chat_1 = require("./net/Chat");
__exportStar(require("./net/CDSI"), exports);
__exportStar(require("./net/Chat"), exports);
// This must match the libsignal-bridge Rust enum of the same name.
var Environment;
(function (Environment) {
    Environment[Environment["Staging"] = 0] = "Staging";
    Environment[Environment["Production"] = 1] = "Production";
})(Environment || (exports.Environment = Environment = {}));
function newNativeHandle(handle) {
    return {
        _nativeHandle: handle,
    };
}
exports.newNativeHandle = newNativeHandle;
/** Low-level async runtime control, mostly just exported for testing. */
class TokioAsyncContext {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    makeCancellable(abortSignal, promise) {
        if (abortSignal !== undefined) {
            const cancellationToken = promise._cancellationToken;
            const cancel = () => {
                Native.TokioAsyncContext_cancel(this, cancellationToken);
            };
            if (abortSignal.aborted) {
                cancel();
            }
            else {
                abortSignal.addEventListener('abort', cancel);
            }
        }
        return promise;
    }
}
exports.TokioAsyncContext = TokioAsyncContext;
/** The "scheme" for Signal TLS proxies. See {@link Net.setProxy()}. */
exports.SIGNAL_TLS_PROXY_SCHEME = 'org.signal.tls';
class Net {
    constructor(options) {
        this.asyncContext = new TokioAsyncContext(Native.TokioAsyncContext_new());
        if (options.localTestServer) {
            this._connectionManager = newNativeHandle(Native.TESTING_ConnectionManager_newLocalOverride(options.userAgent, options.TESTING_localServer_chatPort, options.TESTING_localServer_cdsiPort, options.TESTING_localServer_svr2Port, options.TESTING_localServer_rootCertificateDer));
        }
        else {
            this._connectionManager = newNativeHandle(Native.ConnectionManager_new(options.env, options.userAgent));
        }
    }
    /**
     * Starts the process of connecting to the chat server.
     *
     * If this completes successfully, the next call to {@link #connectAuthenticatedChat} may be able
     * to finish more quickly. If it's incomplete or produces an error, such a call will start from
     * scratch as usual. Only one preconnect is recorded, so there's no point in calling this more
     * than once.
     *
     * @param options additional options to pass through.
     * @param options.abortSignal an {@link AbortSignal} that will cancel the connection attempt.
     */
    preconnectChat(options) {
        return this.asyncContext.makeCancellable(options?.abortSignal, Native.AuthenticatedChatConnection_preconnect(this.asyncContext, this._connectionManager));
    }
    /**
     *
     * Creates a new instance of {@link UnauthenticatedChatConnection}.
     * @param listener the listener for incoming events.
     * @param options additional options to pass through.
     * @param options.abortSignal an {@link AbortSignal} that will cancel the connection attempt.
     * @returns the connected listener, if the connection succeeds.
     */
    async connectUnauthenticatedChat(listener, options) {
        return Chat_1.UnauthenticatedChatConnection.connect(this.asyncContext, this._connectionManager, listener, options);
    }
    /**
     * Creates a new instance of {@link AuthenticatedChatConnection}.
     */
    connectAuthenticatedChat(username, password, receiveStories, listener, options) {
        return Chat_1.AuthenticatedChatConnection.connect(this.asyncContext, this._connectionManager, username, password, receiveStories, listener, options);
    }
    /**
     * Enables/disables IPv6 for all new connections (until changed).
     *
     * The flag is `true` by default.
     */
    setIpv6Enabled(ipv6Enabled) {
        Native.ConnectionManager_set_ipv6_enabled(this._connectionManager, ipv6Enabled);
    }
    /**
     * Enables or disables censorship circumvention for all new connections (until changed).
     *
     * If CC is enabled, *new* connections and services may try additional routes to the Signal
     * servers. Existing connections and services will continue with the setting they were created
     * with. (In particular, changing this setting will not affect any existing
     * {@link ChatConnection ChatConnections}.)
     *
     * CC is off by default.
     */
    setCensorshipCircumventionEnabled(enabled) {
        Native.ConnectionManager_set_censorship_circumvention_enabled(this._connectionManager, enabled);
    }
    setProxy(hostOrOptions, portOrNothing) {
        if (typeof hostOrOptions === 'string') {
            // Support <username>@<host> syntax to allow UNENCRYPTED_FOR_TESTING as a marker user.
            // This is not a stable feature of the API and may go away in the future;
            // the Rust layer will reject any other users anyway. But it's convenient for us.
            const [before, after] = hostOrOptions.split('@', 2);
            const [username, domain] = after ? [before, after] : [undefined, before];
            hostOrOptions = {
                scheme: exports.SIGNAL_TLS_PROXY_SCHEME,
                host: domain,
                port: portOrNothing,
                username,
            };
        }
        const { scheme, host, port, username, password } = hostOrOptions;
        try {
            const proxyConfig = newNativeHandle(Native.ConnectionProxyConfig_new(scheme, host, 
            // i32::MIN represents "no port provided"; we don't expect anyone to pass that manually.
            port ?? -2147483648, username ?? null, password ?? null));
            Native.ConnectionManager_set_proxy(this._connectionManager, proxyConfig);
        }
        catch (e) {
            this.setInvalidProxy();
            throw e;
        }
    }
    /**
     * Like {@link #setProxy}, but parses the proxy options from a URL. See there for more
     * information.
     *
     * Takes a string rather than a URL so that an *invalid* string can result in disabling
     * connections until {@link #clearProxy} is called, consistent with other ways {@link #setProxy}
     * might consider its parameters invalid.
     *
     * Throws if the URL contains unnecessary parts (like a query string), or if the resulting options
     * are not supported.
     */
    setProxyFromUrl(urlString) {
        let options;
        try {
            options = Net.proxyOptionsFromUrl(urlString);
        }
        catch (e) {
            // Make sure we set an invalid proxy on error,
            // so no connection can be made until the problem is fixed.
            this.setInvalidProxy();
            throw e;
        }
        this.setProxy(options);
    }
    /**
     * Parses a proxy URL into an options object, suitable for passing to {@link #setProxy}.
     *
     * It is recommended not to call this directly. Instead, use {@link #setProxyFromUrl}, which will
     * treat an invalid URL uniformly with one that is structurally valid but unsupported by
     * libsignal.
     *
     * Throws if the URL is known to not be a valid proxy URL; however it's still possible the
     * resulting options object cannot be used as a proxy.
     */
    static proxyOptionsFromUrl(urlString) {
        const url = new URL(urlString);
        // Check all the parts of the URL.
        // scheme://username:password@hostname:port/path?query#fragment
        const scheme = url.protocol.slice(0, -1);
        // This does not distinguish between "https://proxy.example" and "https://@proxy.example".
        // This could be done by manually checking `url.href`.
        // But until someone complains about it, let's not worry about it.
        const username = url.username != '' ? url.username : undefined;
        const password = url.password != '' ? url.password : undefined;
        const host = url.hostname;
        const port = url.port != '' ? Number.parseInt(url.port, 10) : undefined;
        if (url.pathname != '' && url.pathname != '/') {
            throw new Error('proxy URLs should not have path components');
        }
        if (url.search != '') {
            throw new Error('proxy URLs should not have query components');
        }
        if (url.hash != '') {
            throw new Error('proxy URLs should not have fragment components');
        }
        return { scheme, username, password, host, port };
    }
    /**
     * Refuses to make any new connections until a new proxy configuration is set or
     * {@link #clearProxy} is called.
     *
     * Existing connections will not be affected.
     */
    setInvalidProxy() {
        Native.ConnectionManager_set_invalid_proxy(this._connectionManager);
    }
    /**
     * Ensures that future connections will be made directly, not through a proxy.
     *
     * Clears any proxy configuration set via {@link #setProxy} or {@link #setInvalidProxy}. If none
     * was set, calling this method is a no-op.
     */
    clearProxy() {
        Native.ConnectionManager_clear_proxy(this._connectionManager);
    }
    /**
     * Notifies libsignal that the network has changed.
     *
     * This will lead to, e.g. caches being cleared and cooldowns being reset.
     */
    onNetworkChange() {
        Native.ConnectionManager_on_network_change(this._connectionManager);
    }
    async cdsiLookup(auth, options) {
        return (0, CDSI_1.cdsiLookup)({
            asyncContext: this.asyncContext,
            connectionManager: this._connectionManager,
        }, auth, options);
    }
}
exports.Net = Net;
