// @signalapp/libsignal-client/dist/zkgroup/internal/ByteArray.js
"use strict";
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNCHECKED_AND_UNCLONED = void 0;
const Errors_1 = require("../../Errors");
const Native = require("../../../Native");
exports.UNCHECKED_AND_UNCLONED = Symbol();
class ByteArray {
    constructor(contents, checkValid) {
        if (checkValid === exports.UNCHECKED_AND_UNCLONED) {
            this.contents = contents;
        }
        else {
            checkValid.call(Native, contents);
            this.contents = Buffer.from(contents);
        }
    }
    static checkLength(expectedLength) {
        return (contents) => {
            if (contents.length !== expectedLength) {
                throw new Errors_1.LibSignalErrorBase(`Length of array supplied was ${contents.length} expected ${expectedLength}`, undefined, this.name);
            }
        };
    }
    getContents() {
        return this.contents;
    }
    serialize() {
        return Buffer.from(this.contents);
    }
}
exports.default = ByteArray;
