// @signalapp/libsignal-client/dist/net/Chat.js
"use strict";
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildHttpRequest = exports.AuthenticatedChatConnection = exports.UnauthenticatedChatConnection = exports.ChatServerMessageAck = void 0;
const Native = require("../../Native");
const node_buffer_1 = require("node:buffer");
const net_1 = require("../net");
const DEFAULT_CHAT_REQUEST_TIMEOUT_MILLIS = 5000;
class ChatServerMessageAck {
    constructor(_nativeHandle) {
        this._nativeHandle = _nativeHandle;
    }
    send(statusCode) {
        Native.ServerMessageAck_SendStatus(this, statusCode);
    }
}
exports.ChatServerMessageAck = ChatServerMessageAck;
class ConnectionInfoImpl {
    constructor(_nativeHandle) {
        this._nativeHandle = _nativeHandle;
    }
    get localPort() {
        return Native.ChatConnectionInfo_local_port(this);
    }
    get ipVersion() {
        const value = Native.ChatConnectionInfo_ip_version(this);
        switch (value) {
            case 1:
                return 'IPv4';
            case 2:
                return 'IPv6';
            default:
                throw new TypeError(`ip type was unexpectedly ${value}`);
        }
    }
    toString() {
        return Native.ChatConnectionInfo_description(this);
    }
}
class UnauthenticatedChatConnection {
    static async connect(asyncContext, connectionManager, listener, options) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const connect = Native.UnauthenticatedChatConnection_connect(asyncContext, connectionManager);
        const chat = await asyncContext.makeCancellable(options?.abortSignal, connect);
        const connection = (0, net_1.newNativeHandle)(chat);
        Native.UnauthenticatedChatConnection_init_listener(connection, new WeakListenerWrapper(nativeChatListener));
        return new UnauthenticatedChatConnection(asyncContext, connection, nativeChatListener);
    }
    /**
     * Creates a chat connection backed by a fake remote end.
     *
     * @param asyncContext the async runtime to use
     * @param listener the listener to send events to
     * @returns an {@link UnauthenticatedChatConnection} and handle for the remote
     * end of the fake connection.
     */
    static fakeConnect(asyncContext, listener) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const fakeChat = (0, net_1.newNativeHandle)(Native.TESTING_FakeChatConnection_Create(asyncContext, new WeakListenerWrapper(nativeChatListener), ''));
        const chat = (0, net_1.newNativeHandle)(Native.TESTING_FakeChatConnection_TakeUnauthenticatedChat(fakeChat));
        return [
            new UnauthenticatedChatConnection(asyncContext, chat, nativeChatListener),
            (0, net_1.newNativeHandle)(Native.TESTING_FakeChatConnection_TakeRemote(fakeChat)),
        ];
    }
    constructor(asyncContext, chatService, 
    // Unused except to keep the listener alive since the Rust code only holds a
    // weak reference to the same object.
    chatListener) {
        this.asyncContext = asyncContext;
        this.chatService = chatService;
        this.chatListener = chatListener;
    }
    fetch(chatRequest, options) {
        return this.asyncContext.makeCancellable(options?.abortSignal, Native.UnauthenticatedChatConnection_send(this.asyncContext, this.chatService, buildHttpRequest(chatRequest), chatRequest.timeoutMillis ?? DEFAULT_CHAT_REQUEST_TIMEOUT_MILLIS));
    }
    disconnect() {
        return Native.UnauthenticatedChatConnection_disconnect(this.asyncContext, this.chatService);
    }
    connectionInfo() {
        return new ConnectionInfoImpl(Native.UnauthenticatedChatConnection_info(this.chatService));
    }
}
exports.UnauthenticatedChatConnection = UnauthenticatedChatConnection;
class AuthenticatedChatConnection {
    static async connect(asyncContext, connectionManager, username, password, receiveStories, listener, options) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const connect = Native.AuthenticatedChatConnection_connect(asyncContext, connectionManager, username, password, receiveStories);
        const chat = await asyncContext.makeCancellable(options?.abortSignal, connect);
        const connection = (0, net_1.newNativeHandle)(chat);
        Native.AuthenticatedChatConnection_init_listener(connection, new WeakListenerWrapper(nativeChatListener));
        return new AuthenticatedChatConnection(asyncContext, connection, nativeChatListener);
    }
    /**
     * Creates a chat connection backed by a fake remote end.
     *
     * @param asyncContext the async runtime to use
     * @param listener the listener to send events to
     * @param alerts alerts to send immediately upon connect
     * @returns an {@link AuthenticatedChatConnection} and handle for the remote
     * end of the fake connection.
     */
    static fakeConnect(asyncContext, listener, alerts) {
        const nativeChatListener = makeNativeChatListener(asyncContext, listener);
        const fakeChat = (0, net_1.newNativeHandle)(Native.TESTING_FakeChatConnection_Create(asyncContext, new WeakListenerWrapper(nativeChatListener), alerts?.join('\n') ?? ''));
        const chat = (0, net_1.newNativeHandle)(Native.TESTING_FakeChatConnection_TakeAuthenticatedChat(fakeChat));
        return [
            new AuthenticatedChatConnection(asyncContext, chat, nativeChatListener),
            (0, net_1.newNativeHandle)(Native.TESTING_FakeChatConnection_TakeRemote(fakeChat)),
        ];
    }
    constructor(asyncContext, chatService, 
    // Unused except to keep the listener alive since the Rust code only holds a
    // weak reference to the same object.
    chatListener) {
        this.asyncContext = asyncContext;
        this.chatService = chatService;
        this.chatListener = chatListener;
    }
    fetch(chatRequest, options) {
        return this.asyncContext.makeCancellable(options?.abortSignal, Native.AuthenticatedChatConnection_send(this.asyncContext, this.chatService, buildHttpRequest(chatRequest), chatRequest.timeoutMillis ?? DEFAULT_CHAT_REQUEST_TIMEOUT_MILLIS));
    }
    disconnect() {
        return Native.AuthenticatedChatConnection_disconnect(this.asyncContext, this.chatService);
    }
    connectionInfo() {
        return new ConnectionInfoImpl(Native.AuthenticatedChatConnection_info(this.chatService));
    }
}
exports.AuthenticatedChatConnection = AuthenticatedChatConnection;
/**
 * Holds a {@link Native.ChatListener} by {@link WeakRef} and delegates
 * `ChatListener` calls to it.
 *
 * This lets us avoid passing anything across the bridge that has a normal
 * (strong) reference to the app-side listener. The danger is that the passed-in
 * listener might gain a reference to the JS connection object; that would
 * result in a reference cycle that Node can't clean up because one of the
 * references is through a Rust `Box`.
 *
 * When constructing a connection, calling code should wrap an app-side listener
 * in this type and pass it across the bridge, then hold its own strong
 * reference to the same listener as a field. This ensures that if there is a
 * reference cycle between the connection and app-side listener, that cycle is
 * visible to the Node runtime, while still ensuring the passed-in listener
 * stays alive as long as the connection does.
 */
class WeakListenerWrapper {
    constructor(listener) {
        this.listener = new WeakRef(listener);
    }
    _connection_interrupted(reason) {
        this.listener.deref()?._connection_interrupted(reason);
    }
    _incoming_message(envelope, timestamp, ack) {
        this.listener.deref()?._incoming_message(envelope, timestamp, ack);
    }
    _queue_empty() {
        this.listener.deref()?._queue_empty();
    }
    _received_alerts(alerts) {
        this.listener.deref()?._received_alerts(alerts);
    }
}
function makeNativeChatListener(asyncContext, listener) {
    if ('onQueueEmpty' in listener) {
        return {
            _incoming_message(envelope, timestamp, ack) {
                listener.onIncomingMessage(envelope, timestamp, new ChatServerMessageAck(ack));
            },
            _queue_empty() {
                listener.onQueueEmpty();
            },
            _received_alerts(alerts) {
                listener.onReceivedAlerts?.(alerts);
            },
            _connection_interrupted(cause) {
                listener.onConnectionInterrupted(cause);
            },
        };
    }
    return {
        _incoming_message(_envelope, _timestamp, _ack) {
            throw new Error('Event not supported on unauthenticated connection');
        },
        _queue_empty() {
            throw new Error('Event not supported on unauthenticated connection');
        },
        _received_alerts(alerts) {
            if (alerts.length != 0) {
                throw new Error(`Got ${alerts.length} unexpected alerts on an unauthenticated connection`);
            }
        },
        _connection_interrupted(cause) {
            listener.onConnectionInterrupted(cause);
        },
    };
}
function buildHttpRequest(chatRequest) {
    const { verb, path, body, headers } = chatRequest;
    const bodyBuffer = body !== undefined ? node_buffer_1.Buffer.from(body) : null;
    const httpRequest = {
        _nativeHandle: Native.HttpRequest_new(verb, path, bodyBuffer),
    };
    headers.forEach((header) => {
        const [name, value] = header;
        Native.HttpRequest_add_header(httpRequest, name, value);
    });
    return httpRequest;
}
exports.buildHttpRequest = buildHttpRequest;
